subsys/power/device.c

```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
 /* Number of devices successfully suspended. */
 static device_idx_t num_susp;
 
+static bool should_suspend(const struct device *dev, uint32_t state)
+{
+	int rc;
+	uint32_t current_state;
+
+	rc = device_get_power_state(dev, &current_state);
+	if ((rc != -ENOTSUP) && (rc != 0)) {
+		LOG_DBG("Was not possible to get device %s state: %d",
+			dev->name, rc);
+		return true;
+	}
+
+	/*
+	 * If the device is currently powered off or the request was
+	 * to go to the same state, just ignore it.
+	 */
+	if ((current_state == DEVICE_PM_OFF_STATE) ||
+			(current_state == state)) {
+		return false;
+	}
+
+	return true;
+}
+
 static int _pm_devices(uint32_t state)
 {
 	num_susp = 0;
 
 	for (int i = num_pm - 1; i >= 0; i--) {
 		device_idx_t idx = pm_devices[i];
 		const struct device *dev = &all_devices[idx];
+		bool suspend;
 		int rc;
 
-		/* TODO: Improve the logic by checking device status
-		 * and set the device states accordingly.
-		 */
-		rc = device_set_power_state(dev, state, NULL, NULL);
-		if ((rc != -ENOTSUP) && (rc != 0)) {
-			LOG_DBG("%s did not enter %s state: %d",
-				dev->name, device_pm_state_str(state), rc);
-			return rc;
+		suspend = should_suspend(dev, state);
+		if (suspend) {
+			/*
+			 * Don't bother the device if it is currently
+			 * in the right state.
+			 */
+			rc = device_set_power_state(dev, state, NULL, NULL);
+			if ((rc != -ENOTSUP) && (rc != 0)) {
+				LOG_DBG("%s did not enter %s state: %d",
+					dev->name, device_pm_state_str(state),
+					rc);
+				return rc;
```

Comment:
in case of error, some devices will be left on `state` while some others will be kept on the current state. I wonder if this function should *undo* the state change in case of error.

Reply:
Nope, the first device that fails will cause this function to return with an error code. The system pm then will undo the changes calling `pm_resume_devices()` that will only iterate over the devices that have changed state.